versioned_docs/version-2021fa/lecture3.md

```diff
@@ -42,4 +42,364 @@ From now on, we will be using a `tsconfig.json` file within every Node project w
+app.get('/posts', (_, res) => {
+  res.send(posts1);
+});
+
+// use firebase instead
+app.get('/getPosts', async (_, res) => {
+  const postsSnapshot = await postsCollection.get();
+  const allPostsDoc = postsSnapshot.docs;
+  const posts: PostWithID[] = [];
+  for (let doc of allPostsDoc) {
+    const post: PostWithID = doc.data() as PostWithID;
+    post.id = doc.id;
+    posts.push(post);
+  }
+  res.send(posts);
+});
+
+// read posts by name
+app.get('/posts/:name', async function (req, res) {
+  const name = req.params.name;
+  const postsSnapshot = await postsCollection.where('name', '==', name).get();
+  const allPostsDoc = postsSnapshot.docs;
+  const posts: PostWithID[] = [];
+  for (let doc of allPostsDoc) {
+    const post: PostWithID = doc.data() as PostWithID;
+    post.id = doc.id;
+    posts.push(post);
+  }
+  res.send(posts);
+});
+
+// read posts by id
+app.get('/postById/:id', async function (req, res) {
+  const id = req.params.id;
+  const postsSnapshot = await postsCollection.doc(id).get();
+  const post: PostWithID = postsSnapshot.data() as PostWithID;
+  res.send(post);
+});
+
+// sort posts by name
+app.get('/postsorted', async function (req, res) {
+  const postsSnapshot = await postsCollection.orderBy('name', 'desc').get();
+  const allPostsDoc = postsSnapshot.docs;
+  const posts: PostWithID[] = [];
+  for (let doc of allPostsDoc) {
+    const post: PostWithID = doc.data() as PostWithID;
+    post.id = doc.id;
+    posts.push(post);
+  }
+  res.send(posts);
+});
+
+// POST method: create a new post
+app.post('/addPost', (req, res) => {
+  const post: Post = req.body;
+  posts1.push(post);
+  res.send(`Post created by ${req.body.name}!`);
+});
+
+// generate a document with a random name to store the post's data
+app.post('/addPost2', async function (req, res) {
+  const post: Post = req.body;
+  const postDoc = postsCollection.doc();
+  await postDoc.set(post);
+  res.send(postDoc.id);
+});
+
+// POST method: update an existing post
+app.post('/updatePost', (req, res) => {
+  for (let post of posts1) {
+    if (post.name === req.body.name) {
+      post.content = req.body.content;
+    }
+  }
+  console.log(posts1);
+  res.send('content updated!');
+});
+
+// update by id
+app.post('/updatePost2/:id', async function (req, res) {
+  const newPost: Post = req.body;
+  const id: string = req.params.id;
+  await postsCollection.doc(id).update(newPost);
+  res.send('updated!');
+});
+
+// DELETE methdod: delete a post
+app.delete('/removePost', (req, res) => {
+  const newPosts = [];
+  for (let post of posts1) {
+    if (post.name !== req.body.name) {
+      newPosts.push(post);
+    }
+  }
+  posts1 = newPosts;
+  res.send(`Post by ${req.body.name} deleted!`);
+});
+
+// delete by id
+app.delete('/removePost/:id', async function (req, res) {
+  const id = req.params.id;
+  await postsCollection.doc(id).delete();
+  res.send('deleted!');
+});
+
+app.listen(port, () => console.log('App started!'));
```

Comment:
Let's make it print the port or a link to `http://localhost:port` in the `"App started!"` message.